Gui: Fix window center rotation mode and orthographic view clipping

Rexbas opened the pull request:
This PR fixes #11243 and re-fixes #6836. #9699, The original fix for #6836 was a huge improvement but I was still able to reproduce the cut corners and sometimes random glitches. With this fix that does not happen anymore, at least during my testing.

The new approach finds a bounding sphere of the scene, then it is repositioning the camera such that the `nearDistance` can be 0 and the `farDistance` the diameter of the bounding sphere. Finally, to fix #11243, we change the focal distance so the focal point stays the same.

Currently, this fix runs only in `NavigationStyle::reorientCamera(..)`. There are also other methods that change the camera where it is still possible to get glitchy results.  I am going to take a look at coin3d `SoRenderManagerP::setClippingPlanes()` to see if I can make a fix there, so this PR is maybe a temporary fix.

wwmayer commented:
> This PR fixes https://github.com/FreeCAD/FreeCAD/issues/11243

I can confirm that it fixes this issue. Btw, a very easy way to reproduce this behaviour is to rotate, stop rotate, zoom in or out with the wheel and then rotate again.

However, I am bit concerned about the way how it is fixed. It determines the scene bounding box and therefore it has to traverse the whole scene graph and this  makes it  a rather expensive operation. Because reorientCamera() is called for every little movement it could affect the speed of user interactions.

Actually it suffices to determine the scene bounding box only once when starting to rotate.

Btw, which previous PR caused this regression?

Rexbas commented:
> Actually it suffices to determine the scene bounding box only once when starting to rotate.

Yeah you are right, I added a commit that finds the bounding sphere before dragging or animating and also after the scene graph is set. The bounding sphere is stored in a private member of `NavigationStyle` and `reorientCamera()` uses this value. Yesterday, I opened a PR (https://github.com/coin3d/coin/pull/510) in coin with almost the same solution as in this PR. If that gets merged into coin then most of this PR could be reverted.

> Btw, which previous PR caused this regression?

PR #10881 caused this regression, specifically commit 9be3bfe7bf7f3f76f896e7cb5175509d5b3c6289. In this commit I changed `reorientCamera()` to be able to handle rotating about a point that is not the focal point, which was needed for animations that combine rotation and translation. However, another part of the problem came from #9699. Here the focal distance was changed depending on the height which is related to zoom but the focal distance affects the rotation center. And the new `reorientCamera()` did not like that.

maxwxyz commented:
Can confirm this fixes the issue of the rotation center moving around.
Although when panning the new rotation center is fairly offset from the last position, shouldn't it stay at the same distance/offset(=depth) when I just pan to the right?

Rexbas commented:
> shouldn't it stay at the same distance/offset(=depth) when I just pan to the right?

Yes, if you align the camera with FRONT (or whatever the nearest orientation is) before panning then the rotation center should stay in the object. In the video your camera is not aligned so it makes sense that the rotation center is moving away from the object.

maxwxyz commented:
So the rotation center moves on a perpendicular plane to the view direction (with an offset of the last position)?

Rexbas commented:
Yes

maxwxyz commented:
is this now a partial fix or a full fix of the mentioned issues/PRs? Does this need to wait until the Coin PR is merged or is this PR finished? From a user PoV this is working.

Rexbas commented:
In my opinion this is a full fix for #11243 and a partial/temporary fix for #6836 to resolve the worst of the clipping problems. We don't have to wait for the PR I created in coin. That is approximately the same fix but in a better place and we can revert most of this PR when we update to a coin version that includes the PR I created in coin.

adrianinsaval commented:
@wwmayer should we still worry about performance here or do you think this one is ok now?

wwmayer commented:
Should be fine now.

Rexbas commented:
Thanks, I can reproduce it and I will try to fix it.

maxwxyz commented:
@Rexbas shall I create an issue or are you on it?

Rexbas commented:
I am currently working on another issue and have planned to do this after it is finished.

Rexbas commented:
@maxwxyz, I looked into it but I don't think this is an issue we can fix completely. First, it does not only happen to the CAD style and window center rotation mode. It happens to all styles and all rotation modes. Also, you don't necessary need a large scene. It happens when you zoom in a lot, which is more likely when you have a large scene.

The reason why we can't completely fix this is because I suspect this is a float precision issue. The camera uses floats for orientation and position and when the camera is rotated there is almost always a small error. This error is small enough to not notice it in most cases. But when you zoom in a lot then the error becomes noticeable in the form of a moving rotation center. But really it is the camera not moving perfectly due to inaccuracies. I tried the same in Fusion 360 and there the rotation center also moves when zoomed in a lot.

So I don't know how to fix this or how to reduce the problem, it is probably not even possible to fix it completely. If you think this is an important problem, consider opening an issue. Maybe someone has an idea. But I don't.
